### PR TITLE
improve performance of static UCR loading

### PR DESCRIPTION
**Pull request by snopoke**
http://manage.dimagi.com/default.asp?229845
@nickpell 

cc @czue

### CODE REVIEW

**Comment by snopoke** [comment]
For the ICDS sync (which has 59 report fixtures) this reduces the sync time from 110s to 11s (when I run it locally).

**Comment by benrudolph** [comment]
😮 😮

File: corehq/apps/userreports/models.py [hunk 1]
```diff
@@ -460,20 +464,31 @@ def get_doc_id(cls, domain, table_id):
         return '{}{}-{}'.format(cls._datasource_id_prefix, domain, table_id)
 
     @classmethod
-    def all(cls):
+    @skippable_quickcache([], skip_arg='rebuild')
```
Comment (benrudolph) [inline corehq/apps/userreports/models.py]: was this the big bread winner?

**Comment by nickpell** [comment]
Test failure looks related:
```
ERROR: test_bootstrap_can_be_called (corehq.apps.userreports.tests.test_pillow.StaticKafkaIndicatorPillowTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/vendor/lib/python2.7/site-packages/mock.py", line 1224, in patched
    return func(*args, **keywargs)
  File "/mnt/commcare-hq-ro/corehq/apps/userreports/tests/test_pillow.py", line 186, in test_bootstrap_can_be_called
    get_kafka_ucr_static_pillow().bootstrap()
  File "/mnt/commcare-hq-ro/corehq/apps/userreports/pillow.py", line 161, in bootstrap
    self._processor.bootstrap(configs)
  File "/mnt/commcare-hq-ro/corehq/apps/userreports/pillow.py", line 51, in bootstrap
    self.rebuild_tables_if_necessary()
  File "/mnt/commcare-hq-ro/corehq/apps/userreports/pillow.py", line 98, in rebuild_tables_if_necessary
    self.rebuild_table(sql_adapter)
  File "/mnt/commcare-hq-ro/corehq/apps/userreports/pillow.py", line 108, in rebuild_table
    rebuild_indicators.delay(sql_adapter.config.get_id)
  File "/vendor/lib/python2.7/site-packages/celery/app/task.py", line 453, in delay
    return self.apply_async(args, kwargs)
  File "/vendor/lib/python2.7/site-packages/celery/app/task.py", line 551, in apply_async
    link=link, link_error=link_error, **options)
  File "/vendor/lib/python2.7/site-packages/celery/app/task.py", line 743, in apply
    request=request, propagate=throw)
  File "/vendor/lib/python2.7/site-packages/celery/app/trace.py", line 355, in eager_trace_task
    uuid, args, kwargs, request)
  File "/vendor/lib/python2.7/site-packages/celery/app/trace.py", line 253, in trace_task
    I, R, state, retval = on_error(task_request, exc, uuid)
  File "/vendor/lib/python2.7/site-packages/celery/app/trace.py", line 240, in trace_task
    R = retval = fun(*args, **kwargs)
  File "/mnt/commcare-hq-ro/corehq/apps/userreports/tasks.py", line 38, in rebuild_indicators
    config = _get_config_by_id(indicator_config_id)
  File "/mnt/commcare-hq-ro/corehq/apps/userreports/tasks.py", line 17, in _get_config_by_id
    return StaticDataSourceConfiguration.by_id(indicator_config_id)
  File "/mnt/commcare-hq-ro/corehq/apps/userreports/models.py", line 517, in by_id
    return cls._get_from_metadata(metadata)
  File "/mnt/commcare-hq-ro/corehq/apps/userreports/models.py", line 523, in _get_from_metadata
    return cls._get_datasource_config(wrapped, metadata.domain)
  File "/mnt/commcare-hq-ro/corehq/apps/userreports/models.py", line 529, in _get_datasource_config
    doc['_id'] = cls.get_doc_id(domain, doc['table_id'])
KeyError: 'table_id'
```

File: corehq/apps/userreports/models.py (hunk 1 above)
Comment (snopoke) [inline corehq/apps/userreports/models.py]: yea, having the map in the first place and caching it so we don't have to regenerate it for each call

**Comment by snopoke** [comment]
Struggling to reproduce test failure locally, will try more tomorrow

**Comment by czue** [comment]
looks good to merge on :white_check_mark: from me. if this is blocking ICDS I'm also fine `skip`ing the test temporarily to chase down the issue

File: corehq/apps/userreports/models.py [hunk 2]
```diff
@@ -519,17 +552,23 @@ def get_doc_id(cls, domain, report_id, custom_configurable_report):
...
+        for wrapped, path in StaticReportConfiguration._all():
+            for domain in wrapped.domains:
+                config_id = cls.get_doc_id(domain, wrapped.report_id, wrapped.custom_configurable_report)
+                mapping[config_id] = StaticReportMetadata(config_id, path, domain)
+        return mapping
 
     @classmethod
     def all(cls):
-        for wrapped in StaticReportConfiguration._all():
+        for wrapped, _ in StaticReportConfiguration._all():
```
Comment (dimagimon) [inline corehq/apps/userreports/models.py]: F402 import '_' from line 39 shadowed by loop variable

File: corehq/apps/userreports/models.py [hunk 3]
```diff
@@ -543,39 +582,58 @@ def by_id(cls, config_id):
...
+                raise ReportConfigurationNotFoundError(_(
+                    "The following report configuration could not be found: {}".format(config_id)
+                ))
+            return_configs.append(cls._get_from_metadata(metadata))
         return return_configs
 
     @classmethod
     def report_class_by_domain_and_id(cls, domain, config_id):
-        for wrapped in cls._all():
+        for wrapped, _ in cls._all():
```
Comment (dimagimon) [inline corehq/apps/userreports/models.py]: F402 import '_' from line 39 shadowed by loop variable

File: corehq/apps/userreports/models.py [hunk 4]
```diff
@@ -460,20 +464,31 @@ def get_doc_id(cls, domain, table_id):
...
         for provider_path in settings.STATIC_DATA_SOURCE_PROVIDERS:
             provider_fn = to_function(provider_path, failhard=True)
-            for datasource in provider_fn():
-                yield datasource
+            for wrapped, path in provider_fn():
+                yield wrapped, path
+
+    @classmethod
+    def all(cls):
+        for wrapped, _ in cls._all():
```
Comment (dimagimon) [inline corehq/apps/userreports/models.py]: F402 import '_' from line 39 shadowed by loop variable

**Comment by snopoke** [comment]
:white_check_mark:

**Pull request merged**